gno.land/pkg/integration/gnoland.go

```diff
@@ -0,0 +1,359 @@
+	"github.com/gnolang/gno/tm2/pkg/crypto"
+	"github.com/gnolang/gno/tm2/pkg/db"
+	"github.com/gnolang/gno/tm2/pkg/log"
+	osm "github.com/gnolang/gno/tm2/pkg/os"
+	"github.com/gnolang/gno/tm2/pkg/std"
+	"github.com/rogpeppe/go-internal/testscript"
+)
+
+type IntegrationConfig struct {
+	SkipFailingGenesisTxs bool
+	SkipStart             bool
+	GenesisBalancesFile   string
+	GenesisTxsFile        string
+	ChainID               string
+	GenesisRemote         string
+	RootDir               string
+	GenesisMaxVMCycles    int64
+	Config                string
+}
+
+// NOTE: This is a copy of gnoland actual flags.
+// XXX: A lot this make no sense for integration.
+func (c *IntegrationConfig) RegisterFlags(fs *flag.FlagSet) {
+	fs.BoolVar(
+		&c.SkipFailingGenesisTxs,
+		"skip-failing-genesis-txs",
+		false,
+		"don't panic when replaying invalid genesis txs",
+	)
+	fs.BoolVar(
+		&c.SkipStart,
+		"skip-start",
+		false,
+		"quit after initialization, don't start the node",
+	)
+
+	fs.StringVar(
+		&c.GenesisBalancesFile,
+		"genesis-balances-file",
+		"./genesis/genesis_balances.txt",
+		"initial distribution file",
+	)
+
+	fs.StringVar(
+		&c.GenesisTxsFile,
+		"genesis-txs-file",
+		"./genesis/genesis_txs.txt",
+		"initial txs to replay",
+	)
+
+	fs.StringVar(
+		&c.ChainID,
+		"chainid",
+		"dev",
+		"the ID of the chain",
+	)
+
+	fs.StringVar(
+		&c.RootDir,
+		"root-dir",
+		"testdir",
+		"directory for config and data",
+	)
+
+	fs.StringVar(
+		&c.GenesisRemote,
+		"genesis-remote",
+		"localhost:26657",
+		"replacement for '%%REMOTE%%' in genesis",
+	)
+
+	fs.Int64Var(
+		&c.GenesisMaxVMCycles,
+		"genesis-max-vm-cycles",
+		10_000_000,
+		"set maximum allowed vm cycles per operation. Zero means no limit.",
+	)
+}
+
+func execTestingGnoland(t *testing.T, logger log.Logger, gnoDataDir, gnoRootDir string, args []string) (*node.Node, error) {
+	t.Helper()
+
+	// Setup start config.
+	icfg := &IntegrationConfig{}
+	{
+		fs := flag.NewFlagSet("start", flag.ExitOnError)
+		icfg.RegisterFlags(fs)
+
+		// Override default value for flags.
+		fs.VisitAll(func(f *flag.Flag) {
+			switch f.Name {
+			case "root-dir":
+				f.DefValue = gnoDataDir
+			case "chainid":
+				f.DefValue = "tendermint_test"
+			case "genesis-balances-file":
+				f.DefValue = filepath.Join(gnoRootDir, "gno.land", "genesis", "genesis_balances.txt")
+			case "genesis-txs-file":
+				f.DefValue = filepath.Join(gnoRootDir, "gno.land", "genesis", "genesis_txs.txt")
+			default:
+				return
+			}
+
+			f.Value.Set(f.DefValue)
+		})
+
+		if err := fs.Parse(args); err != nil {
+			return nil, fmt.Errorf("unable to parse flags: %w", err)
+		}
+	}
+
+	// Setup testing config.
+	cfg := config.TestConfig().SetRootDir(gnoDataDir)
+	{
+		cfg.EnsureDirs()
+		cfg.Consensus.CreateEmptyBlocks = true
+		cfg.Consensus.CreateEmptyBlocksInterval = time.Duration(0)
+		cfg.RPC.ListenAddress = "tcp://127.0.0.1:0"
+		cfg.P2P.ListenAddress = "tcp://127.0.0.1:0"
```

Comment:
Let's set a port here instead of doing the dance of "set port 0 - start the node - check the new port"?

Reply:
beside the fact that this is needed because of concurrent run by `testscript`, I don't like using static ports in tests generally speaking, it avoid running tests concurrently and it also doesn't cost too much effort grabbing port and address from listener, in this case, address is directly guessed by the node and you just have to pick it

Reply:
yeah, I still meant a random port.

on second thought it's fine, the port-guessing code is already done inside of gno (I had to fix code on it once) so it's good not to duplicate it. resolving this.